File: CamStudy/CamStudy/ViewModels/UserSessionViewModel.swift
import FirebaseAuth
import FirebaseFirestore
import Foundation

@MainActor
final class UserSessionViewModel: ObservableObject {
    @Published var profile: UserProfile?
    @Published var errorMessage: String?

    private var listener: ListenerRegistration?
    private let firestoreService = FirestoreService.shared

    var userId: String? {
        Auth.auth().currentUser?.uid
    }

    func startObservingProfile() {
        guard listener == nil, let userId else { return }
        listener = firestoreService.observeProfile(userId: userId) { [weak self] result in
            Task { @MainActor in
                guard let self else { return }
                switch result {
                case .success(let profile):
                    self.profile = profile
                case .failure(let error):
                    self.errorMessage = error.localizedDescription
                }
            }
        }
    }

    func stopObservingProfile() {
        listener?.remove()
        listener = nil
    }
}

File: camstudy_app/lib/viewmodels/user_session_view_model.dart
import 'dart:async';

import 'package:flutter/foundation.dart';

import '../models/user_profile.dart';
import '../services/auth_service.dart';
import '../services/firestore_service.dart';

class UserSessionViewModel extends ChangeNotifier {
  UserProfile? profile;
  String? errorMessage;

  StreamSubscription<UserProfile>? _subscription;
  final _firestoreService = FirestoreService.shared;

  String? get userId => AuthService.shared.currentUser?.uid;

  void startObservingProfile() {
    final id = userId;
    if (id == null || _subscription != null) return;
    _subscription = _firestoreService
        .observeProfile(id)
        .listen(
          (profile) {
            this.profile = profile;
            notifyListeners();
          },
          onError: (Object error) {
            errorMessage = error.toString();
            notifyListeners();
          },
        );
  }

  void stopObservingProfile() {
    _subscription?.cancel();
    _subscription = null;
  }

  @override
  void dispose() {
    stopObservingProfile();
    super.dispose();
  }
}
